### PR TITLE
Envelop Command

### PR DESCRIPTION
**Pull request by samlown**
Adding an `envelop` command which will apply all the build options to the document instead of the envelope.

This implies that stuff like `--set title="foo"` will apply to the document, and not the envelope. I think this is what you'd expect.

Given a source file like:

```yaml
title: "Test Message"
content: |-
  We hope you like this test message!
```

Usage will look like:

```
gobl envelop -t note.Message message.yaml
```

Outputs:

```json
{
	"$schema": "https://gobl.org/draft-0/envelope",
	"head": {
		"uuid": "7a855e69-a44b-11ec-affe-3e7e00ce5635",
		"dig": {
			"alg": "sha256",
			"val": "45ac3115c8569a1789e58af8d0dc91ef3baa1fb71daaf38f5aef94f82b4d0033"
		}
	},
	"doc": {
		"$schema": "https://gobl.org/draft-0/note/message",
		"title": "Test Message",
		"content": "We hope you like this test message!"
	},
	"sigs": [
		"eyJhbGciOiJFUzI1NiIsImtpZCI6IjBhMjg2MDAwLTM2MGEtNGU2Ni04MWFhLTU2ZDQ0YmI4ZjEwNyJ9.eyJ1dWlkIjoiN2E4NTVlNjktYTQ0Yi0xMWVjLWFmZmUtM2U3ZTAwY2U1NjM1IiwiZGlnIjp7ImFsZyI6InNoYTI1NiIsInZhbCI6IjQ1YWMzMTE1Yzg1NjlhMTc4OWU1OGFmOGQwZGM5MWVmM2JhYTFmYjcxZGFhZjM4ZjVhZWY5NGY4MmI0ZDAwMzMifX0.cca8saAPQoIj95-O_7ibCYGN6gsgcjPiTrFJTyG8pWPpMsWUDS9tZhsWEFWL6eL3LOOW3u4GK_1QYtfjmDsi9w"
	]
}
```

### CODE REVIEW

File: cmd/gobl/build_test.go [hunk 1]
```diff
@@ -341,26 +357,30 @@ func Test_build(t *testing.T) {
...
 			if tt.err != "" {
 				assert.EqualError(t, err, tt.err)
 			} else {
 				assert.Nil(t, err)
 			}
-			re := testy.Replacement{
+			if tt.replace == nil {
+				tt.replace = make([]testy.Replacement, 0)
+			}
+			tt.replace = append(tt.replace, testy.Replacement{
```
Comment (flimzy) [inline cmd/gobl/build_test.go:370]: I think if you reverse the order here, you can skip the `make()` call above.

i.e:

```
replace := append(testy.Replacement{ ... }, tt.replace...)
```

File: internal/build.go [hunk 2]
```diff
@@ -81,9 +79,17 @@ func Build(ctx context.Context, opts BuildOptions) (*gobl.Envelope, error) {
 	if err != nil {
 		return nil, err
 	}
-	env := new(gobl.Envelope)
-	if err = json.Unmarshal(encoded, env); err != nil {
-		return nil, echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	var env *gobl.Envelope
+	if opts.Envelop {
```
Comment (flimzy) [inline internal/build.go:83]: The fact that we're checking this same flag in several places makes me nervous. This pattern makes for tightly coupled code, and is very error prone.

I'd much rather have a `Build()` and `Envelop()` method respectively, that then call unexported methods for shared functionality.

Comment (samlown) [reply internal/build.go:83]: My feeling was that you'd end up with loads of duplicated code. If you'd like to refactor this so it fits your proposed style, please do so! I was most interested in a quick fix to test out the "feel" of the "envelop" command. I accomplished that, and still like it, so please feel free to make it better :-) I'm current focussed on conversations and documentation, so it'll take me a bit more effort to come back to this!

File: cmd/gobl/build.go [hunk 3]
```diff
@@ -28,25 +30,37 @@ func build() *buildOpts {
...
 		Args: cobra.MaximumNArgs(2),
 		RunE: b.runE,
 	}
+	if b.envelop {
+		cmd.Use = "envelop [infile] [outfile]"
+		cmd.Short = "Prepare a document and insert into a new envelope"
+	} else {
+		cmd.Use = "build [infile] [outfile]"
+		cmd.Short = "Combine and complete envelope data"
+	}
```
[the remark this answers is not in the document]
Comment (samlown) [reply cmd/gobl/build.go:48]: maybe... but with all the types you're talking many multiples of extra lines for the sake of one `if`! I've refactored these options into the `build()` and `envelope()` functions which I think addresses your concerns.

File: internal/build.go (hunk 2 above)
Comment (samlown) [reply internal/build.go]: I've given it a go. It's a bit better, although I'm not 100% convinced :-)

File: cmd/gobl/build.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (flimzy) [reply cmd/gobl/build.go:55]: Now that it looks like it's a single `if`, I'm much less concerned. It's the fact of spreading identical logic around in multiple places that bothers me. 

https://www.youtube.com/watch?v=z43bmaMwagI if you're interested in an entertaining take on the issue.

**Review by flimzy — APPROVED** [review]

File: cmd/gobl/build.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (samlown) [reply cmd/gobl/build.go:55]: hehe watching. I actually spent an hour (maybe more) in the first attempt debugging the fact that the `if` statement was the wrong way round, so I'll no doubt learn some more :-)

**Pull request merged**